### PR TITLE
Document environment declarations and related functionality

### PR DESCRIPTION
**Pull request by parlough**
We've needed to document this for a long time and we would like to link to it from the Flutter site as well. 
 
Closes https://github.com/dart-lang/site-www/issues/4349

### CODE REVIEW

**Comment by parlough** [comment]
@lrhn Do you think you could take a look at this? It's just a rough start but I've never used `fromEnvironment` so I want to make sure I'm understanding everything correctly.

**Staged:** https://dart-dev--pr4575-feature-environment-aautlydq.web.app/guides/environment-declarations

File: examples/misc/lib/development/environment_declarations.dart [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+void log(String message) {
+  // Log the debug message if the environment declaration 'DEBUG' is `true`.
+  // If there was no value specified, do not log.
+  if (const bool.fromEnvironment('DEBUG', defaultValue: false)) {
+    print('Debug: $message');
+  }
+}
+// #enddocregion debug-log
+
+void hasDebug() {
```
Comment (lrhn) [inline examples/misc/lib/development/environment_declarations.dart:11]: Should be a getter.

File: src/_guides/environment-declarations.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
+---
+title: Configuring apps with environment declarations
+description: >
+  Learn about using environment declarations to customize application behavior.
+---
+
+Dart supports specifying environment declarations when
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:7]: Consider calling it a "compilation environment".

The word "environment" is overloaded, and traditionally refers to the OS environment, which is available as `Platform.environment`.

The naming of these constructors as `fromEnvironment` has been a constant pain since day 1, so anything we can do to distinguish their meaning from "the other environment" (aka "the real environment") is positive.

File: src/_guides/environment-declarations.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+building or running an application.
+Environment declarations are used to specify
+configuration values or options as key-value pairs
+that are accessed and evaluated at compile time.
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:11]: The "at compile time" is why calling this a "compilation environment" should make sense to the user.

File: src/_guides/environment-declarations.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+Your app can use the values of environment declarations
+to change its functionality or behavior.
+Dart compilers can take advantage of this to eliminate code
+no longer reached due to the values specified through tree shaking.
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:16]: Comma before "through", otherwise it sounds like the values are specified through tree shaking.

Consider quoting "tree shaking", readers are not necessarily expected to know what it means.

File: src/_guides/environment-declarations.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+{{site.alert.note}}
+  The environment declaration constructors are only guaranteed
+  to work when invoked as `const`.
+  Most compilers must be able to evaluate their value at compile time.
+{{site.alert.end}}
+
+Each of the `fromEnvironment` constructors require the
+name or key of the environment declaration key-value pair,
+as well as an optional `defaultValue` override
+that is used when a declaration wasn't defined.
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:60]: , or the value string wasn't valid as an `int` or `bool` result.

(The default value is used for `int.fromEnv...` when the value string cannot be parsed as an integer, and for `bool.fromEnv...` when the value isn't `true` or `false`.)

File: src/_guides/environment-declarations.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+For example, if you only want to log messages only
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:62]: Two "only"s.

File: src/_guides/environment-declarations.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  // Log the debug message if the environment declaration 'DEBUG' is `true`.
+  // If there was no value specified, do not log.
+  if (const bool.fromEnvironment('DEBUG', defaultValue: false)) {
+    print('Debug: $message');
+  }
+}
+```
+
+In this snippet, if `DEBUG` is set to `false`
+during compilation or not specified at all,
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:77]: Comma before "or"?

File: src/_guides/environment-declarations.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+production compilers can completely remove the condition and its body.
+
+The `fromEnvironment` constructors always
+have a default value when the declaration isn't specified.
+Therefore, if you need to specifically check if
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:82]: last "if" -> "whether"?

File: src/_guides/environment-declarations.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+{{site.alert.warning}}
+  Dart tools and compilers currently do not
+  consistently handle environment declarations
+  with comma-separated values.
+  For more details,
+  see [SDK issue 44995][].
+{{site.alert.end}}
+
+[SDK issue 44995]: https://github.com/dart-lang/sdk/issues/4499
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:108]: Link looks wrong, missing "5" at end.

File: src/_guides/environment-declarations.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+$ dart compile jit-snapshot --define=DEBUG=true -DFLAVOR=free main.dart
+$ dart compile kernel --define=DEBUG=true -DFLAVOR=free main.dart
+```
+
+To specify multiple declarations, 
+you can use multiple define options:
+
+```terminal
+$ dart run --define=DEBUG=true -DFLAVOR=free
+```
```
Comment (lrhn) [inline src/_guides/environment-declarations.md:130]: This last section feels spurious after having six lines above showing the same thing already.

**Review by lrhn — APPROVED** [review]
LGTM, only a few nits, and one big request: Consider changing "environment" to "compilation environment" *everywhere*.

We really need to drill into user's minds that this is not the same as the platform/OS environment, and establish that there *exists* a separate compilation environment.

File: src/_guides/environment-declarations.md (hunk 2 above)
Comment (parlough) [reply src/_guides/environment-declarations.md:7]: Thanks for all the great feedback! This is the last remaining suggestion which I'm still working on.

It makes a lot of sense to make it clear these are tied to a "compilation environment", but "compilation environment declarations" is rather long to call these multiple times across the page (and site). So I'm going to give some more thought to how to adjust each mention.

Comment (MaryaBelanger) [reply src/_guides/environment-declarations.md:7]: I would change the title to "Configuring apps with _compilation_ environment declarations" and clarify the distinction right in the first paragraph with an extra sentence.

> Dart supports specifying compilation environment declarations when
> building or running an application.
> Compilation environment declarations allow you to specify
> configuration values or options as key-value pairs
> that are accessed and evaluated at compile time.
> This page uses "environment" and "environment declarations"
> to refer to the **compilation environment**, not the OS environment
> traditionally referred to as "the environment" in software terminology.

(or "colloquially" instead of "traditionally" in the last line)

Comment (parlough) [reply src/_guides/environment-declarations.md]: I made those adjustments, instead putting the extra sentence as a note. Thanks for the idea :)

**Review by MaryaBelanger — APPROVED** [review]
👍

File: src/_guides/environment-declarations.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+with `const` or within a constant context.
+Use [`bool.fromEnvironment`][bool-from] for `true` or `false` values,
+[`int.fromEnvironment`][int-from] for integer values,
+and [`String.fromEnvironment`][string-from] for anything else.
+
+{{site.alert.note}}
+  The environment declaration constructors are only guaranteed
+  to work when invoked as `const`.
+  Most compilers must be able to evaluate their value at compile time.
+{{site.alert.end}}
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:61]: **question:** Is this not redundant with the previous paragraph?

File: src/_guides/environment-declarations.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+Each of the `fromEnvironment` constructors require the
+name or key of the environment declaration key-value pair,
+as well as an optional `defaultValue` override
+that is used when a declaration wasn't defined
+or the specified value cannot be parsed as the expected type.
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:67]: **issue:** This should be broken up. It's a 42-word run-on sentence.

File: src/_guides/environment-declarations.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+For example, if you only want to log messages
+when the environment declaration `DEBUG` is set to `true`:
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:70]: **suggestion:** Rephrase this to be active voice:
```suggestion
**For example:** If you want to log messages
only when you set the `DEBUG` environment declaration to `true`:
```

File: src/_guides/environment-declarations.md [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+  // If there was no value specified, do not log.
+  if (const bool.fromEnvironment('DEBUG', defaultValue: false)) {
+    print('Debug: $message');
+  }
+}
+```
+
+In this snippet, if `DEBUG` is set to `false`
+during compilation, or not specified at all,
+production compilers can completely remove the condition and its body.
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:85]: **suggestion:** Same as last note.

File: src/_guides/environment-declarations.md [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+
+The `fromEnvironment` constructors always
+have a default value for when the declaration isn't specified
+or the specified value cannot be parsed.
+Therefore, if you need to specifically check whether
+an environment declaration has been specified,
+use the [`bool.hasEnvironment`][bool-has] constructor:
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:92]: **issue:** Cut these sentences up. They're too long.

File: src/_guides/environment-declarations.md [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+[bool-has]: {{site.dart-api}}/{{site.data.pkg-vers.SDK.channel}}/dart-core/bool/bool.hasEnvironment.html
+
+## Specifying environment declarations
+
+{{site.alert.warning}}
+  Dart tools and compilers currently do not
+  consistently handle environment declarations
+  with comma-separated values.
+  For more details,
+  see [SDK issue 44995][].
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:113]: **suggestion:** This could read better. Some additional information from #44495 may make this be more direct.

File: src/_guides/environment-declarations.md [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+$ dart compile aot-snapshot --define=DEBUG=true -DFLAVOR=free main.dart
+$ dart compile jit-snapshot --define=DEBUG=true -DFLAVOR=free main.dart
+$ dart compile kernel --define=DEBUG=true -DFLAVOR=free main.dart
+```
+
+#### `webdev`
+
+To learn about configuring `webdev` to pass environment declarations
+to both the development and production web compilers,
+see [the `webdev` configuration documentation][webdev-config].
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:137]: **chore:** Change `see` to `check out`. It's more inclusive language.

File: src/_guides/environment-declarations.md [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+        "request": "launch",
+        "type": "dart",
+        "toolArgs": [
+          "--define=DEBUG=true"
+        ]
+    }
+]
+```
+
+For more information, see the documentation for
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:159]: **chore:** Change `see` to `check out`. It's more inclusive language. Also, replace `For more information` to `To learn more`.

File: src/_guides/environment-declarations.md [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+[VSC instructions]: https://code.visualstudio.com/docs/editor/debugging#_launch-configurations
+
+### JetBrains IDEs
+
+In the **Run/Debug Configurations** for your project,
+add your desired environment declarations to **VM options**:
+
+![Adding define option to Jetbrains IDE](/assets/img/env-decl-jetbrains.png){:width="500"}
+
+For more information, see the JetBrains' documentation for
```
Comment (atsansone) [inline src/_guides/environment-declarations.md:171]: **chore:** Change `see` to `check out`. It's more inclusive language. Also, replace `For more information` to `To learn more`.

**Review by atsansone — CHANGES_REQUESTED** [review]
@parlough : Apologies for the delay. I offered some suggestions to improve readability. Please address anything marked as **issue** or **chore**. Others you can ignore.

File: src/_guides/environment-declarations.md (hunk 11 above)
Comment (parlough) [reply src/_guides/environment-declarations.md:61]: The previous paragraph does go into similar details, but technically it's possible to try to call them not in a `const` context (even if you shouldn't). This has caused a lot of confusion and problems in the past, so I think it's worth it to be extra explicit here.

File: src/_guides/environment-declarations.md (hunk 13 above)
Comment (parlough) [reply src/_guides/environment-declarations.md:71]: Thinking about this, I feel there isn't a great subject/actor to use (nor is it important), since it may not be the developer themself setting the declaration, it may a user of their package, a CI system, etc.

Let me know what you think.

File: src/_guides/environment-declarations.md (hunk 15 above)
Comment (parlough) [reply src/_guides/environment-declarations.md:93]: I tried a few different ways of breaking them up, but most things I tried didn't read as well. I made a few other modifications to the text, but let me know if you have any ideas.

File: src/_guides/environment-declarations.md [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+Dart supports specifying compilation environment declarations when
+building or running an application.
+Compilation environment declarations are used to specify
+configuration values or options as key-value pairs
+that are accessed and evaluated at compile time.
+
+{{site.alert.note}}
+  This page uses "environment" and "environment declarations"
+  to refer to the compilation environment, not the OS environment
+  often referred to as "the environment" in software terminology.
```
[the remark this answers is not in the document]
Comment (parlough) [reply src/_guides/environment-declarations.md:16]: I've made some other adjustments here, please take a look.

File: src/_guides/environment-declarations.md (hunk 14 above)
Comment (parlough) [reply src/_guides/environment-declarations.md:86]: Same comment as last response.

File: src/_guides/environment-declarations.md [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+{{site.alert.end}}
+
+Your app can use the values of environment declarations
+to change its functionality or behavior.
+Dart compilers can take advantage of this to eliminate code
+no longer reachable due to the values specified, through "tree shaking".
```
[the remark this answers is not in the document]
Comment (parlough) [reply src/_guides/environment-declarations.md:22]: This comment isn't about changing the environment declaration values, but rather using them to make changes.

Also, the second sentence is "tree shaking", so I'm not sure this form of separation makes sense. I've decided to remove "tree shaking", since it doesn't really add anything unless we have a link destination for the term. 

I've pushed some changes, particularly for the second sentence to make it more clear when code can be removed. I continued to use the word "eliminate" though as that is the standard term used in dead code elimination.

File: src/_guides/environment-declarations.md [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+To specify environment declarations to the Flutter tool,
+use the `--dart-define` option instead:
+
+```terminal
+$ flutter run ---dart-define=DEBUG=true
+```
+
+{%- comment %}
+  TODO: Once Flutter adds `--dart-define` documentation:
+  To learn more, check out Flutter's documentation on `--dart define`.
```
Comment (domesticmouse) [inline src/_guides/environment-declarations.md:193]: Missing `-` in `--dart define`?

**Review by domesticmouse — APPROVED** [review]

File: src/_guides/environment-declarations.md (hunk 22 above)
Comment (parlough) [reply src/_guides/environment-declarations.md]: Good catch, thanks.

**Review by atsansone — APPROVED** [review]
LGTM. Thanks!
